crypto/lhash/lhash.c

```diff
@@ -156,7 +156,26 @@ void *OPENSSL_LH_retrieve(OPENSSL_LHASH *lh, const void *data)
     unsigned long hash;
     OPENSSL_LH_NODE **rn;
     void *ret;
+#ifdef HAVE_ATOMICS
+    CRYPTO_REF_TYPE junk = 0;
+# ifdef __GNUC__
+    int nil = 0;
 
+    __atomic_store(&lh->error, &nil, __ATOMIC_RELEASE);
+# else
+    atomic_store(&lh->error, 0);
+# endif
+
+    rn = getrn(lh, data, &hash);
+
+    if (*rn == NULL) {
+        CRYPTO_UP_REF(&lh->num_retrieve_miss, &junk, NULL);
```

Comment:
You mean something that wouldn't require junk variable?

Reply:
Yes. It can be a simple call to atomic_fetch_add_explicit()

Reply:
I'd like to first hear if this would be an alternative to #6754. But yes, I can meanwhile think of all this... I mean do provide feedback, it's just that I'll address it later :-)

Reply:
Perhaps @mingtaoy can try this PR and see if it fixes the errors?

Reply:
That would be just swell...

Reply:
And what if I can confirm it myself? I mean with @mingtaoy's example program. I actually did try, and it illuminated that one needs three more atomic increments (in getrn). And on top of that it even complained about x509v3_cache_extensions, about initial `if (x->ex_flags & EXFLAG_SET) return;` For reference, that's how it looks after #6162:
```
    if (x->ex_flags & EXFLAG_SET)
        return;

    CRYPTO_THREAD_write_lock(x->lock);
    if (x->ex_flags & EXFLAG_SET) {
        CRYPTO_THREAD_unlock(x->lock);
        return;
    }
```
Note code duplication...

But in either case. If I fix all this, test program reports no problems. Now question is if we would agree that it's the way to go. I.e. as alternative to #6754. Or in other words if I should proceed...

Reply:
A bit off-topic, as for x509v3_cache_extensions. One can argue that it would be appropriate to resolve it by promoting EXFLAG_SET to structure field of it own. So that one can perform atomic load on that. Basically it complains about conflicts with **other** bit-operations.

Reply:
proceed.  I will close #6754.  I would like you to think about deleting that first paragraph in the NOTE section (about ANSI C and function pointers).

Reply:
Sorry, was off the grid yesterday.

I can confirm that with this PR, the tsan error for lh_retrieve goes away as expected since all of the writes are now atomics.

The other tsan related error, as @dot-asm pointed out, is the incorrect double checked locking implementation in x509v3_cache_extensions. I was going to report this in a separate issue.

@dot-asm - Although it would definitely be more "clean" to have the set bit as a separate field, I don't think it is strictly necessary. I was thinking something like:

```
int flags = CRYPTO_atomic_read(x->ex_flags); // acquire
if (flags & EXFLAG_SET)
   return;
CRYPTO_THREAD_write_lock(x->lock);
flags = CRYPTO_atomic_read(x->ex_flags); // this can be relaxed
 if (flags & EXFLAG_SET) {
        CRYPTO_THREAD_unlock(x->lock);
        return;
 }
```

Reply:
Also, for context on all of this: 

I have been recently looking into the x509_verify_cert code path because we have been encountering sporadic, hard-to-reproduce certificate verification failures on random hosts. All of the certificates look good (and can be verified manually), so I was suspecting that there might be some data-race in the verification code path.

Another suspicious code path that I was looking into is the initial `X509_NAME_cmp` performed at the start of `X509_check_issued` (`X509_NAME_cmp` does some lazy computation on building the canonical encoding of the X509 structure, and performs a write), although I have not yet been able to reproduce this yet.

Reply:
```
int flags = CRYPTO_atomic_read(x->ex_flags); // acquire
if (flags & EXFLAG_SET)
   return;
```
Nope, doesn't work. In sense the tsan complains. It should be possible to use atomic_fetch_or (and friends), but it would be just too hairy...

Reply:
> I wonder if we should make something other than CRYPTO_UP_REF for this. It's not really a reference counter, we only use the up function. It's also doing more than what we need.

It seems this is a perennial topic; I remember at least https://github.com/openssl/openssl/pull/4549#discussion_r145534717 but think there were other instances as well.

Reply:
> It seems this is a perennial topic;

Well, it might have to be till it's actually fixed. In sense that everything that was said in the context of lhash.c applies even to referred request. More specifically that it is/was overkill for intended purpose. Well, one can say that on most popular platform it isn't that bad, but question if this is the kind of attitude (favouring one platform) that got us to where we are. So with this in mind "fixed" would be to find better balance...

Reply:
> we have been encountering sporadic, irreproducible certificate verification failures on random hosts.

If you think about it, it's not actually impossible to imagine that it can be attributed to that another thing in v3_purp.c. Indeed, as you set EXFLAG_SET at the end of that function, it can become visible to another thread, but without guarantee that preceding modifications are visible. Formally there should be a barrier prior EXFLAG_SET. Such barrier is effectively guaranteed by CRYPTO_THREAD_unlock. So one can wonder if moving EXLFAG_SET past unlock would be appropriate. Naturally provided that we want to keep the "fast-track" check in the beginning. This of course wouldn't impress tsan, only separate structure field would. Otherwise removal of the "fast-track" check would be only appropriate. Or should one say "meanwhile"?